### PR TITLE
|MessageLoopImpl::FlushTasks| runs one task at a time

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

The FlushTasks runs all the remaining tasks in a single batch, which means they are always running on the same thread. It will be invalid if one of the invocation (not the last one) changes the thread configuration.
Running the task before dequeuing the next task will solve this problem.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/64562 as a result

## Tests

I added the following tests:

FollowingTasksSwitchQueueIfFirstTaskMergesThreads

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: fml/message_loop_impl.cc [hunk 1]
```diff
@@ -119,17 +119,22 @@ void MessageLoopImpl::DoTerminate() {
 
 void MessageLoopImpl::FlushTasks(FlushType type) {
   TRACE_EVENT0("fml", "MessageLoop::FlushTasks");
-  std::vector<fml::closure> invocations;
 
-  task_queue_->GetTasksToRunNow(queue_id_, type, invocations);
-
-  for (const auto& invocation : invocations) {
+  const auto now = fml::TimePoint::Now();
+  for (;;) {
```
Comment (iskakaushik) [inline fml/message_loop_impl.cc:124]: is there a way to change this to `While(TasksAvailableOnQueue(queue_id_))`?

File: fml/message_loop_task_queues_merge_unmerge_unittests.cc [hunk 2]
```diff
@@ -159,10 +164,23 @@ TEST(MessageLoopTaskQueueMergeUnmerge,
...
+    fml::closure invocation = task_queue->GetNextTaskToRun(queue_id_1, now);
+    if (!invocation) {
+      break;
+    }
+  }
+
   latch_1.Wait();
 
-  task_queue->GetTasksToRunNow(queue_id_2, fml::FlushType::kAll, invocations);
+  for (;;) {
```
Comment (iskakaushik) [inline fml/message_loop_task_queues_merge_unmerge_unittests.cc:177]: create a helper method in this test to invoke this. ITs done in multiple places.

File: fml/message_loop_impl.cc (hunk 1 above)
Comment (cyanglaz) [reply fml/message_loop_impl.cc:124]: Yeah, we can create a new method to just return a bool if task is available, and another method to return the actual task. Although I think that might be redundant. 
What about make `GetNextTaskToRun` return a bool, and also takes a pointer so the invocation is set as well?
For example `bool GetNextTaskToRun(TaskQueueId queue_id, fml:closure& invocation)` ?

Comment (blasten) [reply fml/message_loop_impl.cc:124]: +1

File: fml/message_loop_impl.cc [hunk 3]
```diff
@@ -119,17 +119,22 @@ void MessageLoopImpl::DoTerminate() {
 
 void MessageLoopImpl::FlushTasks(FlushType type) {
```
Comment (blasten) [inline fml/message_loop_impl.cc:120]: are there any unit tests for `FlushTasks`?

Comment (cyanglaz) [reply fml/message_loop_impl.cc:120]: `FlushTasks` is private so there weren't unit tests for it.

Comment (blasten) [reply fml/message_loop_impl.cc:120]: what about the public API `RunExpiredTasksNow` or `RunSingleExpiredTaskNow`?  I think we can repro the bug by just calling those, no?

Comment (cyanglaz) [reply fml/message_loop_impl.cc:120]: Yes you are right. `FollowingTasksSwitchQueueIfFirstTaskMergesThreads` also catches the bug but testing `RunExpiredTasksNow` would be better. I will see to set it up.

File: fml/raster_thread_merger_unittests.cc [hunk 4]
```diff
@@ -301,3 +304,93 @@ TEST(RasterThreadMerger, HandleTaskQueuesAreTheSame) {
...
+    loop2->Run();
+  });
+
+  latch1.Wait();
+  latch2.Wait();
+
+  fml::TaskQueueId qid1 = loop1->GetTaskRunner()->GetTaskQueueId();
+  fml::TaskQueueId qid2 = loop2->GetTaskRunner()->GetTaskQueueId();
+  const auto raster_thread_merger_ =
+      fml::MakeRefCounted<fml::RasterThreadMerger>(qid1, qid2);
```
Comment (blasten) [inline fml/raster_thread_merger_unittests.cc:333]: minor nit: maybe these can be named raster and platform queues, so it's easier to follow?

File: fml/raster_thread_merger_unittests.cc [hunk 5]
```diff
@@ -301,3 +304,93 @@ TEST(RasterThreadMerger, HandleTaskQueuesAreTheSame) {
   term1.Signal();
   thread1.join();
 }
+
+TEST(RasterThreadMerger, FirstTaskMergesAndUnmergesThreads) {
```
Comment (blasten) [inline fml/raster_thread_merger_unittests.cc:308]: nice!

File: fml/message_loop_impl.cc (hunk 3 above)
Comment (cyanglaz) [reply fml/message_loop_impl.cc:120]: done

File: fml/message_loop_task_queues_merge_unmerge_unittests.cc (hunk 2 above)
Comment (cyanglaz) [reply fml/message_loop_task_queues_merge_unmerge_unittests.cc]: done

File: fml/message_loop_impl.cc (hunk 1 above)
Comment (cyanglaz) [reply fml/message_loop_impl.cc]: done

File: fml/raster_thread_merger_unittests.cc (hunk 4 above)
Comment (cyanglaz) [reply fml/raster_thread_merger_unittests.cc]: done

File: fml/raster_thread_merger_unittests.cc (hunk 5 above)
Comment (cyanglaz) [reply fml/raster_thread_merger_unittests.cc]: Separated this test into 2 tests since you reviewed last time.

**Pull request merged**